### PR TITLE
feat: classes for common secret types

### LINKED ISSUE #344: Provide a simple API for common types of secrets
https://kubernetes.io/docs/concepts/configuration/secret/#docker-config-secrets
https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret
https://kubernetes.io/docs/concepts/configuration/secret/#ssh-authentication-secrets
https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets
https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/

### PR DESCRIPTION
**Pull request by Chriscbr**
This PR adds custom classes for the following secret types:

- service account token secrets
- basic authentication secrets
- tls secrets
- ssh authentication secrets
- docker config secrets

Closes #344

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

### CODE REVIEW

File: test/secret.test.ts [hunk 1]
```diff
@@ -65,3 +65,100 @@ Array [
 ]
 `);
 });
+
```
Comment (iliapolo) [inline test/secret.test.ts:68]: Missing two tests?

- BasicAuth
- SshAuth

File: test/secret.test.ts [hunk 2]
```diff
@@ -65,3 +65,100 @@ Array [
...
+test('Can create an ssh authentication secret', () => {
```
Comment (iliapolo) [inline test/secret.test.ts]: I think the name here is wrong...

```suggestion
test('Can create a service account token secret', () => {
```

File: test/secret.test.ts [hunk 3]
```diff
@@ -65,3 +65,100 @@ Array [
...
+  const chart = Testing.chart();
+
+  const sa = new kplus.ServiceAccount(chart, 'ServiceAccount');
+  const secret = new kplus.ServiceAccountTokenSecret(chart, 'BasicAuth', {
```
Comment (iliapolo) [inline test/secret.test.ts]:
```suggestion
  const secret = new kplus.ServiceAccountTokenSecret(chart, 'ServiceAccountToken', {
```

File: src/secret.ts [hunk 4]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+   * @see https://docs.docker.com/engine/reference/commandline/cli/#sample-configuration-file
+   */
+  readonly data: { [key: string]: any };
+}
+
+/**
+ * Create a secret for storing credentials for accessing a container image
+ * registry.
+ *
+ * @see https://docs.docker.com/engine/reference/commandline/cli/#sample-configuration-file
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
 * @see https://kubernetes.io/docs/concepts/configuration/secret/#docker-config-secrets
```

File: src/secret.ts [hunk 5]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+  readonly tlsCert: string;
+
+  /**
+   * The TLS key
+   */
+  readonly tlsKey: string;
+}
+
+/**
+ * Create a secret for storing a certificate and its associated key.
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
 * Create a secret for storing a TLS certificate and its associated key.
 *
 * @see https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets
```

File: src/secret.ts [hunk 6]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+ */
+export interface ServiceAccountTokenSecretProps extends ResourceProps {
+  /**
+   * The service account to store a secret for
+   */
+  readonly serviceAccount: IServiceAccount;
+}
+
+/**
+ * Create a secret for basic authentication.
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
 * Create a secret for a service account token.
 *
 * @see https://kubernetes.io/docs/concepts/configuration/secret/#service-account-token-secrets
```

File: src/secret.ts [hunk 7]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+ */
+export interface SshAuthSecretProps extends ResourceProps {
+  /**
+   * The SSH credential to use
+   */
+  readonly sshPrivateKey: string;
+}
+
+/**
+ * Create a secret for basic authentication.
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
 * Create a secret for ssh authentication.
 *
 * @see https://kubernetes.io/docs/concepts/configuration/secret/#ssh-authentication-secrets
```

File: src/secret.ts [hunk 8]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+  readonly username: string;
+
+  /**
+   * The password or token for authentication
+   */
+  readonly password: string;
+}
+
+/**
+ * Create a secret for basic authentication.
```
Comment (iliapolo) [inline src/secret.ts:118]:
```suggestion
 * Create a secret for basic authentication.
 *
 * @see https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret
```

**Review by iliapolo — CHANGES_REQUESTED** [review]

File: src/secret.ts [hunk 9]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+  readonly serviceAccount: IServiceAccount;
+}
+
+/**
+ * Create a secret for basic authentication.
+ */
+export class ServiceAccountTokenSecret extends Secret {
+  public constructor(scope: Construct, id: string, props: ServiceAccountTokenSecretProps) {
+    super(scope, id, {
+      type: 'kubernetes.io/basic-auth',
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
      type: 'kubernetes.io/service-account-token',
```

File: src/secret.ts [hunk 10]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+  readonly sshPrivateKey: string;
+}
+
+/**
+ * Create a secret for basic authentication.
+ */
+export class SshAuthSecret extends Secret {
+  public constructor(scope: Construct, id: string, props: SshAuthSecretProps) {
+    super(scope, id, {
+      type: 'kubernetes.io/basic-auth',
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
      type: 'kubernetes.io/ssh-auth',
```

File: src/secret.ts [hunk 11]
```diff
@@ -94,3 +98,143 @@ export class Secret extends Resource implements ISecret {
...
+    });
+  }
+}
+
+/**
+ * Options for `SshAuthSecret`.
+ */
+export interface SshAuthSecretProps extends ResourceProps {
+  /**
+   * The SSH credential to use
```
Comment (iliapolo) [inline src/secret.ts]:
```suggestion
   * The SSH private key to use
```

**Review by iliapolo — CHANGES_REQUESTED** [review]

**Comment by Chriscbr** [comment]
Ah I forgot committing from the GitHub UI messes up all the DCO checks -- I'm just going to squash and force push all of these commits.

File: test/secret.test.ts (hunk 1 above)
Comment (Chriscbr) [reply test/secret.test.ts:68]: Fixed

**Review by iliapolo — APPROVED** [review]

**Comment by iliapolo** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-21/main|[](https://github.com/cdk8s-team/cdk8s-plus/pull/565)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by iliapolo** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-20/main|[](https://github.com/cdk8s-team/cdk8s-plus/pull/568)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**